src/systemctl/sysv-compat.c

```diff
@@ -65,38 +65,65 @@ int parse_shutdown_time_spec(const char *t, usec_t *ret) {
 
                 *ret = now(CLOCK_REALTIME) + USEC_PER_MINUTE * u;
         } else {
-                char *e = NULL;
-                long hour, minute;
+                char hour_s[6];
+                char min_s[3];
+                char *colon_position;
+                unsigned long hour, minute;
                 struct tm tm = {};
                 time_t s;
                 usec_t n;
+                bool relative = false;
+                int err;
 
-                errno = 0;
-                hour = strtol(t, &e, 10);
-                if (errno > 0 || *e != ':' || hour < 0 || hour > 23)
+                if (t[0] == '+') {
+                        relative = true;
+                        t++;
+                }
+
+                colon_position = strchr(t,':');
+                if (colon_position == NULL)
+                        return -EINVAL;
+
+                strncpy(hour_s, t, colon_position - t);
+                hour_s[5] = 0;
+
+                strncpy(min_s, colon_position + 1, 2);
+                min_s[2] = 0;
+
+                err = safe_atolu(hour_s,&hour);
+                if (err != 0 || hour > 99999)
                         return -EINVAL;
 
-                minute = strtol(e+1, &e, 10);
-                if (errno > 0 || *e != 0 || minute < 0 || minute > 59)
+                err = safe_atolu(min_s,&minute);
+                if (err != 0 || minute > 59)
                         return -EINVAL;
 
                 n = now(CLOCK_REALTIME);
-                s = (time_t) (n / USEC_PER_SEC);
 
-                assert_se(localtime_r(&s, &tm));
+                if (relative) {
+                        s = USEC_PER_HOUR * hour + USEC_PER_MINUTE * minute;
```

Comment:
something is off here: `s` is time_t, which indicates it stores seconds. but you use USEC_PER_HOUR and friends, which means you deal with µs.

Reply:
This has been altered to use a new usec_t named us.